mysys/my_addr_resolve.c

```diff
@@ -297,48 +304,50 @@ static int addr_resolve(void *ptr, my_addr_loc *loc)
 
 int my_addr_resolve(void *ptr, my_addr_loc *loc)
 {
-  Dl_info info;
+#ifndef __PIE__
+  addr_offset= (void*) info.dli_fbase;
+  if (strcmp(info.dli_fname, my_progname) == 0 &&
+      addr_resolve((void*) my_addr_resolve, loc) == 0 &&
+      strcmp(loc->func, "my_addr_resolve") == 0)
+    addr_offset= 0;
+#endif
 
+  return addr_resolve((void*) (ptr - addr_offset), loc);
+}
+
+
+const char *my_addr_resolve_init(void *ptr)
+{
   if (!dladdr(ptr, &info))
-    return 1;
+    return "dladdr() error";
 
-  if (strcmp(addr2line_binary, info.dli_fname))
```

Comment:
As I understood it correctly, in the original code addr2line_binary was used for two purposes: a) to prevent fork for addr2line launching if it was already forked for the current shared object, b) launch addr2line for each shared object in backtrace. In the new version start_addr2line_fork() is invoked only from my_addr_resolve_init(), i.e. once per the whole backtrace. Suppose our backtrace contains function addresses from different shared libraries. Will it work correctly?